<a href="https://colab.research.google.com/github/nedokormysh/Stepik_Karpov_ML/blob/main/Stepik_group_agg.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Edu

In [1]:
import pandas as pd
import numpy as np

In [2]:
students_performance = pd.read_csv('https://stepik.org/media/attachments/course/4852/StudentsPerformance.csv')
students_performance.columns = [x.replace(" ", "_") for x in students_performance.columns]

In [3]:
students_performance.head()

gender race/ethnicity  ... reading_score writing_score
0  female        group B  ...            72            74
1  female        group C  ...            90            88
2  female        group B  ...            95            93
3    male        group A  ...            57            44
4    male        group C  ...            78            75

[5 rows x 8 columns]

In [4]:
students_performance.groupby('gender')

In [5]:
students_performance.groupby('gender').head(2)

gender race/ethnicity  ... reading_score writing_score
0  female        group B  ...            72            74
1  female        group C  ...            90            88
3    male        group A  ...            57            44
4    male        group C  ...            78            75

[4 rows x 8 columns]

In [6]:
students_performance.groupby('race/ethnicity').mean()

math_score  reading_score  writing_score
race/ethnicity                                          
group A          61.629213      64.674157      62.674157
group B          63.452632      67.352632      65.600000
group C          64.463950      69.103448      67.827586
group D          67.362595      70.030534      70.145038
group E          73.821429      73.028571      71.407143

Чтобы увидеть надпись лейбл чего и название метрики

In [7]:
students_performance.groupby('gender').agg({'math_score': ['mean']})

math_score
             mean
gender           
female  63.633205
male    68.728216

In [8]:
students_performance.groupby('gender').aggregate({"math_score":'mean'})

math_score
gender            
female   63.633205
male     68.728216

Чтобы вывести сразу несколько метрик

In [9]:
students_performance.groupby('gender', as_index=False)\
  .aggregate({'math_score': ['mean', 'count', 'std'],
              'reading_score': ['std', 'min', 'max']})

gender math_score                  reading_score         
                mean count        std           std min  max
0  female  63.633205   518  15.491453     14.378245  17  100
1    male  68.728216   482  14.356277     13.931832  23  100

Хочу попробовать сгруппировать по race. Но по определённому значению, по группе B.

In [10]:
students_performance.groupby(students_performance['race/ethnicity'] == 'group B').mean()

math_score  reading_score  writing_score
race/ethnicity                                          
False            66.707407      69.595062       68.62963
True             63.452632      67.352632       65.60000

In [11]:
students_performance[students_performance['race/ethnicity'] == 'group B'].head()

gender race/ethnicity  ... reading_score writing_score
0  female        group B  ...            72            74
2  female        group B  ...            95            93
5  female        group B  ...            83            78
6  female        group B  ...            95            92
7    male        group B  ...            43            39

[5 rows x 8 columns]

Всё таки сначала нужно отфильтровать по группе, а потом группировать.

In [12]:
students_performance[students_performance['race/ethnicity'] == 'group B'].groupby('race/ethnicity').mean()

math_score  reading_score  writing_score
race/ethnicity                                          
group B          63.452632      67.352632           65.6

In [13]:
students_performance.groupby('gender').mean()

math_score  reading_score  writing_score
gender                                          
female   63.633205      72.608108      72.467181
male     68.728216      65.473029      63.311203

In [14]:
students_performance.groupby('gender').agg({'math_score': 'mean', 'reading_score': 'mean'})

math_score  reading_score
gender                           
female   63.633205      72.608108
male     68.728216      65.473029

In [15]:
students_performance.groupby('gender', as_index=False).agg({'math_score': 'mean', 'reading_score': 'mean'})

gender  math_score  reading_score
0  female   63.633205      72.608108
1    male   68.728216      65.473029

Переименнование переменных.

In [16]:
students_performance.groupby('gender', as_index=False)\
  .agg({'math_score': 'mean', 'reading_score': 'mean'})\
  .rename(columns={'math_score': 'mean_math_score', 'reading_score': 'mean_reading_score'})

gender  mean_math_score  mean_reading_score
0  female        63.633205           72.608108
1    male        68.728216           65.473029

In [17]:
students_performance.groupby('gender', as_index=False)\
  .agg(mean_math_score=('math_score', 'mean'), mean_reading_score = ('reading_score', 'mean'))

gender  mean_math_score  mean_reading_score
0  female        63.633205           72.608108
1    male        68.728216           65.473029

группировка по нескольким переменным сразу

In [18]:
students_performance.groupby(['gender', 'race/ethnicity'], as_index=False)\
  .agg({'math_score': 'mean', 'reading_score': 'mean'})\
  .rename(columns={'math_score': 'mean_math_score', 'reading_score': 'mean_reading_score'})

gender race/ethnicity  mean_math_score  mean_reading_score
0  female        group A        58.527778           69.000000
1  female        group B        61.403846           71.076923
2  female        group C        62.033333           71.944444
3  female        group D        65.248062           74.046512
4  female        group E        70.811594           75.840580
5    male        group A        63.735849           61.735849
6    male        group B        65.930233           62.848837
7    male        group C        67.611511           65.424460
8    male        group D        69.413534           66.135338
9    male        group E        76.746479           70.295775

In [19]:
mean_scores = students_performance.groupby(['gender', 'race/ethnicity'])\
  .agg({'math_score': 'mean', 'reading_score': 'mean'})\
  .rename(columns={'math_score': 'mean_math_score', 'reading_score': 'mean_reading_score'})

mean_scores

mean_math_score  mean_reading_score
gender race/ethnicity                                     
female group A               58.527778           69.000000
       group B               61.403846           71.076923
       group C               62.033333           71.944444
       group D               65.248062           74.046512
       group E               70.811594           75.840580
male   group A               63.735849           61.735849
       group B               65.930233           62.848837
       group C               67.611511           65.424460
       group D               69.413534           66.135338
       group E               76.746479           70.295775

In [20]:
mean_scores.index

MultiIndex([('female', 'group A'),
            ('female', 'group B'),
            ('female', 'group C'),
            ('female', 'group D'),
            ('female', 'group E'),
            (  'male', 'group A'),
            (  'male', 'group B'),
            (  'male', 'group C'),
            (  'male', 'group D'),
            (  'male', 'group E')],
           names=['gender', 'race/ethnicity'])

In [21]:
mean_scores.loc[('female', 'group C')]

mean_math_score       62.033333
mean_reading_score    71.944444
Name: (female, group C), dtype: float64

In [22]:
mean_scores.loc[[('female', 'group C'), ('female', 'group A')]]

mean_math_score  mean_reading_score
gender race/ethnicity                                     
female group C               62.033333           71.944444
       group A               58.527778           69.000000

In [23]:
mean_scores['mean_math_score']

gender  race/ethnicity
female  group A           58.527778
        group B           61.403846
        group C           62.033333
        group D           65.248062
        group E           70.811594
male    group A           63.735849
        group B           65.930233
        group C           67.611511
        group D           69.413534
        group E           76.746479
Name: mean_math_score, dtype: float64

Наши колонки - серии. И к ним можно применять разные крутые методы.

Все уникальные значения.

In [24]:
students_performance.math_score.unique()

array([ 72,  69,  90,  47,  76,  71,  88,  40,  64,  38,  58,  65,  78,
        50,  18,  46,  54,  66,  44,  74,  73,  67,  70,  62,  63,  56,
        97,  81,  75,  57,  55,  53,  59,  82,  77,  33,  52,   0,  79,
        39,  45,  60,  61,  41,  49,  30,  80,  42,  27,  43,  68,  85,
        98,  87,  51,  99,  84,  91,  83,  89,  22, 100,  96,  94,  48,
        35,  34,  86,  92,  37,  28,  24,  26,  95,  36,  29,  32,  93,
        19,  23,   8])

Число уникальных значений

In [25]:
students_performance.math_score.nunique()

81

Число уникальных значений на пересечении двух групп. Группируем по двум переменным. Затем обращаемся к нашей колонке. Получаем серию.

In [26]:
students_performance.groupby(['gender', 'race/ethnicity']).math_score

In [27]:
students_performance.groupby(['gender', 'race/ethnicity']).math_score.nunique()

gender  race/ethnicity
female  group A           29
        group B           51
        group C           59
        group D           53
        group E           44
male    group A           38
        group B           43
        group C           56
        group D           49
        group E           38
Name: math_score, dtype: int64

In [28]:
type(students_performance.groupby(['gender', 'race/ethnicity']).math_score.nunique())

pandas.core.series.Series

Групбай с другими методами. Отберём по пять топ студентов по математике по гендеру.

In [29]:
students_performance.sort_values(['gender', 'math_score'])

gender race/ethnicity  ... reading_score writing_score
59   female        group C  ...            17            10
980  female        group B  ...            24            23
17   female        group B  ...            32            28
787  female        group B  ...            38            32
145  female        group C  ...            39            33
..      ...            ...  ...           ...           ...
306    male        group E  ...            87            81
149    male        group E  ...           100            93
623    male        group A  ...            96            86
625    male        group D  ...            97            99
916    male        group E  ...           100           100

[1000 rows x 8 columns]

In [30]:
students_performance.sort_values(['gender', 'math_score'], ascending=False)

gender race/ethnicity  ... reading_score writing_score
149    male        group E  ...           100            93
623    male        group A  ...            96            86
625    male        group D  ...            97            99
916    male        group E  ...           100           100
306    male        group E  ...            87            81
..      ...            ...  ...           ...           ...
145  female        group C  ...            39            33
787  female        group B  ...            38            32
17   female        group B  ...            32            28
980  female        group B  ...            24            23
59   female        group C  ...            17            10

[1000 rows x 8 columns]

In [31]:
students_performance.sort_values(['gender', 'math_score'], ascending=False).groupby('gender').head()

gender race/ethnicity  ... reading_score writing_score
149    male        group E  ...           100            93
623    male        group A  ...            96            86
625    male        group D  ...            97            99
916    male        group E  ...           100           100
306    male        group E  ...            87            81
451  female        group E  ...            92            97
458  female        group E  ...           100           100
962  female        group E  ...           100           100
114  female        group E  ...           100           100
263  female        group E  ...            93            90

[10 rows x 8 columns]

Добавление данных

In [32]:
students_performance.math_score + students_performance.reading_score + students_performance.writing_score

0      218
1      247
2      278
3      148
4      229
      ... 
995    282
996    172
997    195
998    223
999    249
Length: 1000, dtype: int64

просуммировать можно по-другому. Отфильтровали по скору колонки и сложили

In [44]:
students_performance.filter(like = 'score').sum(axis=1)

0      441.384495
1      499.509388
2      561.627621
3      300.997212
4      463.433722
          ...    
995    569.641907
996    349.147494
997    395.273000
998    451.407172
999    503.517453
Length: 1000, dtype: float64

In [42]:
students_performance['total_score'] = students_performance.math_score + students_performance.reading_score + students_performance.writing_score

In [43]:
students_performance.head()

gender race/ethnicity  ... total_score total_score_log
0  female        group B  ...         218        5.384495
1  female        group C  ...         247        5.509388
2  female        group B  ...         278        5.627621
3    male        group A  ...         148        4.997212
4    male        group C  ...         229        5.433722

[5 rows x 10 columns]

Это простое добавление одной колонки. Но нужно сделать множественное преобразование - assign. Кроме того assign создаёт копию датафрейма.

In [35]:
students_performance.assign(total_score_log = np.log(students_performance.total_score))

gender race/ethnicity  ... total_score total_score_log
0    female        group B  ...         218        5.384495
1    female        group C  ...         247        5.509388
2    female        group B  ...         278        5.627621
3      male        group A  ...         148        4.997212
4      male        group C  ...         229        5.433722
..      ...            ...  ...         ...             ...
995  female        group E  ...         282        5.641907
996    male        group C  ...         172        5.147494
997  female        group C  ...         195        5.273000
998  female        group D  ...         223        5.407172
999  female        group D  ...         249        5.517453

[1000 rows x 10 columns]

удаление

In [36]:
students_performance.drop(['total_score'], axis=1)

gender race/ethnicity  ... reading_score writing_score
0    female        group B  ...            72            74
1    female        group C  ...            90            88
2    female        group B  ...            95            93
3      male        group A  ...            57            44
4      male        group C  ...            78            75
..      ...            ...  ...           ...           ...
995  female        group E  ...            99            95
996    male        group C  ...            55            55
997  female        group C  ...            71            65
998  female        group D  ...            78            77
999  female        group D  ...            86            86

[1000 rows x 8 columns]

Ассайн создаёт копию датафрейма, поэтому ничего не произошло.

In [37]:
students_performance = students_performance.assign(total_score_log = np.log(students_performance.total_score))

drop не изменяет исходный датафрейм, то есть чтобы получить новый датафрейм без колонки lunch нам нужно или явно присвоить результат новому датафрейму:

In [38]:
students_performance.drop(['total_score'], axis=1)

gender race/ethnicity  ... writing_score total_score_log
0    female        group B  ...            74        5.384495
1    female        group C  ...            88        5.509388
2    female        group B  ...            93        5.627621
3      male        group A  ...            44        4.997212
4      male        group C  ...            75        5.433722
..      ...            ...  ...           ...             ...
995  female        group E  ...            95        5.641907
996    male        group C  ...            55        5.147494
997  female        group C  ...            65        5.273000
998  female        group D  ...            77        5.407172
999  female        group D  ...            86        5.517453

[1000 rows x 9 columns]

assign можно в скобках указывать множество новых полей

#test

**1**

In [46]:
df = pd.read_csv('https://stepik.org/media/attachments/course/4852/dota_hero_stats.csv')
df.head()

Unnamed: 0  ...                                              roles
0           0  ...                       ['Carry', 'Escape', 'Nuker']
1           1  ...    ['Initiator', 'Durable', 'Disabler', 'Jungler']
2           2  ...        ['Support', 'Disabler', 'Nuker', 'Durable']
3           3  ...  ['Carry', 'Disabler', 'Jungler', 'Nuker', 'Ini...
4           4  ...        ['Support', 'Disabler', 'Nuker', 'Jungler']

[5 rows x 8 columns]

In [53]:
# df.count(axis=1).head(10)

In [56]:
df.groupby('legs').agg('count')

Unnamed: 0  attack_type  id  localized_name  name  primary_attr  roles
legs                                                                        
0             11           11  11              11    11            11     11
2             95           95  95              95    95            95     95
4              7            7   7               7     7             7      7
6              3            3   3               3     3             3      3
8              1            1   1               1     1             1      1

In [67]:
df.groupby('legs').agg({'legs': ['count']})

legs
     count
legs      
0       11
2       95
4        7
6        3
8        1

In [57]:
df.groupby('legs').size()

legs
0    11
2    95
4     7
6     3
8     1
dtype: int64

In [58]:
df.aggregate({'legs': 'value_counts'}).sort_index()

legs
0    11
2    95
4     7
6     3
8     1

In [63]:
df.legs.value_counts().sort_index()

0    11
2    95
4     7
6     3
8     1
Name: legs, dtype: int64

In [60]:
from pandas.core.algorithms import value_counts
? value_counts

**2**

In [68]:
df = pd.read_csv('https://stepik.org/media/attachments/course/4852/accountancy.csv')
df.head()

Unnamed: 0 Executor Type  Salary
0           0     Pupa    D      63
1           1     Pupa    A     158
2           2     Pupa    D     194
3           3     Pupa    E     109
4           4    Loopa    E     184

In [71]:
df.groupby(['Type', 'Executor']).agg({'Salary': ['mean']}).unstack()

Salary            
                mean            
Executor       Loopa        Pupa
Type                            
A          58.000000  160.833333
B         145.166667   77.000000
C         154.333333   74.500000
D         137.714286  146.500000
E         164.000000  131.200000
F         238.000000  136.250000

**3**

In [72]:
df = pd.read_csv('https://stepik.org/media/attachments/course/4852/dota_hero_stats.csv')
df.head()

Unnamed: 0  ...                                              roles
0           0  ...                       ['Carry', 'Escape', 'Nuker']
1           1  ...    ['Initiator', 'Durable', 'Disabler', 'Jungler']
2           2  ...        ['Support', 'Disabler', 'Nuker', 'Durable']
3           3  ...  ['Carry', 'Disabler', 'Jungler', 'Nuker', 'Ini...
4           4  ...        ['Support', 'Disabler', 'Nuker', 'Jungler']

[5 rows x 8 columns]

In [78]:
# df.describe()

In [73]:
df.groupby(['attack_type', 'primary_attr']).agg('count')

Unnamed: 0  id  legs  localized_name  name  roles
attack_type primary_attr                                                   
Melee       agi                   19  19    19              19    19     19
            int                    2   2     2               2     2      2
            str                   35  35    35              35    35     35
Ranged      agi                   18  18    18              18    18     18
            int                   40  40    40              40    40     40
            str                    3   3     3               3     3      3

In [87]:
df.groupby(['attack_type', 'primary_attr']).agg({'id':'count'}).rename(columns={'id': 'count'}).sort_values('count', ascending=False)

count
attack_type primary_attr       
Ranged      int              40
Melee       str              35
            agi              19
Ranged      agi              18
            str               3
Melee       int               2

In [76]:
df.groupby(['attack_type', 'primary_attr']).size()

attack_type  primary_attr
Melee        agi             19
             int              2
             str             35
Ranged       agi             18
             int             40
             str              3
dtype: int64

In [86]:
df.filter(items=['attack_type','primary_attr']).mode()

attack_type primary_attr
0      Ranged          int

**4**

Аспирант Ростислав изучает метаболом водорослей и получил такую табличку. В ней он записал вид каждой водоросли, её род (группа, объединяющая близкие виды), группа (ещё одно объединение водорослей в крупные фракции) и концентрации анализируемых веществ.

Помогите Ростиславу найти среднюю концентрацию каждого из веществ в каждом из родов (колонка genus)! Для этого проведите группировку датафрэйма, сохранённого в переменной concentrations, и примените метод, сохранив результат в переменной mean_concentrations.

In [89]:
concentrations = pd.read_csv('http://stepik.org/media/attachments/course/4852/algae.csv')
concentrations.head()

species       genus  group  ...   citrate   glucose  oleic_acid
0    Fucus_vesiculosus       Fucus  brown  ...  5.004262  2.548459    6.405165
1  Saccharina_japonica  Saccharina  brown  ...  5.621499  6.013219    4.156700
2       Fucus_serratus       Fucus  brown  ...  1.216767  3.623225    0.304573
3      Fucus_distichus       Fucus  brown  ...  6.423606  2.272724    3.393203
4    Cladophora_fracta  Cladophora  green  ...  6.940588  2.316955    2.528886

[5 rows x 8 columns]

In [96]:
mean_concentrations = concentrations.groupby('genus').agg({'sucrose': 'mean', 'alanin': 'mean', 'citrate': 'mean',\
                                                            'glucose': 'mean', 'oleic_acid': 'mean'})
mean_concentrations

sucrose    alanin   citrate   glucose  oleic_acid
genus                                                          
Ascophyllum  6.825467  0.875429  5.253527  3.414961    2.432526
Cladophora   4.008792  3.997055  5.288311  2.800276    2.263472
Fucus        4.355112  3.566411  4.214878  2.814803    3.367647
Palmaria     0.704580  3.176440  5.573905  3.242090    2.245538
Saccharina   4.183596  3.524207  3.346710  4.980594    4.487252

In [98]:
mean_concentrations = concentrations.groupby('genus').mean()
mean_concentrations

sucrose    alanin   citrate   glucose  oleic_acid
genus                                                          
Ascophyllum  6.825467  0.875429  5.253527  3.414961    2.432526
Cladophora   4.008792  3.997055  5.288311  2.800276    2.263472
Fucus        4.355112  3.566411  4.214878  2.814803    3.367647
Palmaria     0.704580  3.176440  5.573905  3.242090    2.245538
Saccharina   4.183596  3.524207  3.346710  4.980594    4.487252

In [99]:
mean_concentrations = concentrations.groupby('genus').agg(['mean'])
mean_concentrations

sucrose    alanin   citrate   glucose oleic_acid
                 mean      mean      mean      mean       mean
genus                                                         
Ascophyllum  6.825467  0.875429  5.253527  3.414961   2.432526
Cladophora   4.008792  3.997055  5.288311  2.800276   2.263472
Fucus        4.355112  3.566411  4.214878  2.814803   3.367647
Palmaria     0.704580  3.176440  5.573905  3.242090   2.245538
Saccharina   4.183596  3.524207  3.346710  4.980594   4.487252

**5**

Пользуясь предыдущими данными, укажите через пробел (без запятых) чему равны минимальная, средняя и максимальная концентрации аланина (alanin) среди видов рода Fucus. Округлите до 2-ого знака, десятичным разделителем является точка.

Формат ответа:
0.55 6.77 7.48

In [113]:
concentrations.loc[concentrations['genus'] == 'Fucus'].describe()

sucrose    alanin   citrate   glucose  oleic_acid
count  3.000000  3.000000  3.000000  3.000000    3.000000
mean   4.355112  3.566411  4.214878  2.814803    3.367647
std    2.110570  3.149943  2.691679  0.713559    3.050376
min    3.001472  0.346431  1.216767  2.272724    0.304573
25%    3.139171  2.028965  3.110515  2.410591    1.848888
50%    3.276870  3.711498  5.004262  2.548459    3.393203
75%    5.031933  5.176401  5.713934  3.085842    4.899184
max    6.786996  6.641303  6.423606  3.623225    6.405165

In [111]:
concentrations.groupby('genus').agg(['min', 'mean', 'max'])

sucrose                      ... oleic_acid                    
                  min      mean       max  ...        min      mean       max
genus                                      ...                               
Ascophyllum  6.825467  6.825467  6.825467  ...   2.432526  2.432526  2.432526
Cladophora   2.452623  4.008792  5.712284  ...   1.432514  2.263472  2.829016
Fucus        3.001472  4.355112  6.786996  ...   0.304573  3.367647  6.405165
Palmaria     0.704580  0.704580  0.704580  ...   2.245538  2.245538  2.245538
Saccharina   1.636122  4.183596  6.731070  ...   4.156700  4.487252  4.817804

[5 rows x 15 columns]

In [110]:
concentrations.groupby('genus').agg(['min', 'mean', 'max']).loc['Fucus', 'alanin'].round(2)

min     0.35
mean    3.57
max     6.64
Name: Fucus, dtype: float64

In [108]:
np.around(concentrations.query("genus == 'Fucus'").alanin.describe().loc[['min', 'mean', 'max']].values, decimals=2)

array([0.35, 3.57, 6.64])

In [112]:
concentrations.query('genus == "Fucus"').describe()

sucrose    alanin   citrate   glucose  oleic_acid
count  3.000000  3.000000  3.000000  3.000000    3.000000
mean   4.355112  3.566411  4.214878  2.814803    3.367647
std    2.110570  3.149943  2.691679  0.713559    3.050376
min    3.001472  0.346431  1.216767  2.272724    0.304573
25%    3.139171  2.028965  3.110515  2.410591    1.848888
50%    3.276870  3.711498  5.004262  2.548459    3.393203
75%    5.031933  5.176401  5.713934  3.085842    4.899184
max    6.786996  6.641303  6.423606  3.623225    6.405165

In [114]:
concentrations.query("genus == 'Fucus'").aggregate({'alanin': ['max','mean','min'] })

alanin
max   6.641303
mean  3.566411
min   0.346431

нашли всех fucus

In [115]:
concentrations.loc[concentrations['genus'] == 'Fucus']

species  genus  group  ...   citrate   glucose  oleic_acid
0  Fucus_vesiculosus  Fucus  brown  ...  5.004262  2.548459    6.405165
2     Fucus_serratus  Fucus  brown  ...  1.216767  3.623225    0.304573
3    Fucus_distichus  Fucus  brown  ...  6.423606  2.272724    3.393203

[3 rows x 8 columns]

аггрегируем по аланину

In [118]:
concentrations.loc[concentrations['genus'] == 'Fucus'].agg({'alanin': [np.min, np.mean, np.max ]}).round(2)

alanin
amin    0.35
mean    3.57
amax    6.64

In [120]:
concentrations.aggregate({'alanin':['min','mean','max']})

alanin
min   0.346431
mean  3.379068
max   6.881024

**6**

Сгруппируйте данные по переменной group и соотнесите вопросы с ответами

In [121]:
df = pd.read_csv('http://stepik.org/media/attachments/course/4852/algae.csv')
df.head()

species       genus  group  ...   citrate   glucose  oleic_acid
0    Fucus_vesiculosus       Fucus  brown  ...  5.004262  2.548459    6.405165
1  Saccharina_japonica  Saccharina  brown  ...  5.621499  6.013219    4.156700
2       Fucus_serratus       Fucus  brown  ...  1.216767  3.623225    0.304573
3      Fucus_distichus       Fucus  brown  ...  6.423606  2.272724    3.393203
4    Cladophora_fracta  Cladophora  green  ...  6.940588  2.316955    2.528886

[5 rows x 8 columns]

In [128]:
df.groupby('group').describe()

sucrose                      ... oleic_acid                    
        count      mean       std  ...        50%       75%       max
group                              ...                               
brown     6.0  4.709666  2.336471  ...   3.774952  4.652528  6.405165
green     3.0  4.008792  1.634817  ...   2.528886  2.678951  2.829016
red       1.0  0.704580       NaN  ...   2.245538  2.245538  2.245538

[3 rows x 40 columns]

In [132]:
# concentrations.groupby('group').agg({'citrate': 'max'})

In [130]:
df.groupby('group') \
    .agg({'citrate': 'var', 'sucrose': lambda x: x.max() - x.min(), 'glucose': 'count'}) \
    .round(2)

citrate  sucrose  glucose
group                           
brown     5.47     5.19        6
green     3.95     3.26        3
red        NaN     0.00        1

дисперсия концентрации цитрата в бурых, зелёных и красных водорослях
5.47, 3.95, NaN

Размах (максимум - минимум) сахарозы в бурых, зелёных и красных водорослях
5.19, 3.26, 0

Число видов бурых, зелёных и красных водорослей
6, 3, 1

In [131]:
df.groupby('group').aggregate(['count', 'var', np.ptp])

sucrose                      ... oleic_acid                    
        count       var       ptp  ...      count       var       ptp
group                              ...                               
brown       6  5.459098  5.189345  ...          6  4.385233  6.100592
green       3  2.672626  3.259662  ...          3  0.540388  1.396502
red         1       NaN  0.000000  ...          1       NaN  0.000000

[3 rows x 15 columns]